### PR TITLE
Feature/ball placement play

### LINKED ISSUE #950: Ball Placement Play
All robots should be standing still except the robot doing ball placement. 

Play is finished if the ball is within the error margin of the desired area, or our ball placement is canceleld

### PR DESCRIPTION
**Pull request by RobotJente**
closes #1001

### CODE REVIEW

File: src/ApplicationManager.cpp [hunk 1]
```diff
@@ -230,6 +232,7 @@ void ApplicationManager::decidePlay(world_new::World *_world) {
 
     // A new play will be chosen if the current play is not valid to keep, or the roles are all finished, in which case the
     // play is considered finished
+    RTT_SUCCESS(!currentPlay)
```
Comment (RobotJente) [inline src/ApplicationManager.cpp]:
```suggestion
```

File: src/stp/Play.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ void Play::update() noexcept {
         // Make sure we don't re assign with too many robots
         if (world->getWorld()->getUs().size() > stp::control_constants::MAX_ROBOT_COUNT) {
             RTT_ERROR("More robots than ROBOT_COUNT(), aborting update on Play")
-            // Make sure the stpInfos is cleared to trigger a reassign whenever
+            // Make surupdate the stpInfos is cleared to trigger a reassign whenever
```
Comment (RobotJente) [inline src/stp/Play.cpp]:
```suggestion
            // Make sure the stpInfos is cleared to trigger a reassign whenever
```

File: src/stp/invariants/HaltGameStateInvariant.cpp [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 namespace rtt::ai::stp::invariant {
     bool HaltGameStateInvariant::checkInvariant(world_new::view::WorldDataView world, const Field *field) const noexcept {
+        std::cout << "meow" << GameStateManager::getCurrentGameState().getRuleSet().maxRobotVel << std::endl;
+
```
Comment (RobotJente) [inline src/stp/invariants/HaltGameStateInvariant.cpp]:
```suggestion
```

File: src/stp/invariants/WeHaveBallInvariant.cpp [hunk 4]
```diff
@@ -7,6 +7,6 @@
 namespace rtt::ai::stp::invariant {
     bool WeHaveBallInvariant::checkInvariant(world_new::view::WorldDataView world, const Field *field) const noexcept {
         auto& us = world.getUs();
-        return std::any_of(us.begin(), us.end(), [](auto& robot){return robot.hasBall();});
+        return std::any_of(us.begin(), us.end(), [](auto& robot){return robot.hasBall() || true;});
```
Comment (RobotJente) [inline src/stp/invariants/WeHaveBallInvariant.cpp]:
```suggestion
        return std::any_of(us.begin(), us.end(), [](auto& robot){return robot.hasBall();});
```

File: src/stp/new_plays/BallPlacement.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
+//
+// Created by jessevw on 24.03.20.
+//
+
+#include "stp/new_plays/Halt.h"
+#include <stp/invariants/HaltGameStateInvariant.h>
+#include <include/roboteam_ai/stp/invariants/WeHaveBallInvariant.h>
+#include <include/roboteam_ai/stp/new_roles/BallPlacer.h>
```
Comment (timovdk) [inline src/stp/new_plays/BallPlacement.cpp]:
```suggestion
#include <stp/invariants/WeHaveBallInvariant.h>
#include <stp/new_roles/BallPlacer.h>
```

File: src/ApplicationManager.cpp [hunk 6]
```diff
@@ -16,6 +16,7 @@
 #include "stp/new_plays/Defend.h"
 #include "stp/new_plays/Attack.h"
 #include <stp/new_plays/Halt.h>
+#include <include/roboteam_ai/stp/new_plays/BallPlacement.h>
```
Comment (RobotJente) [inline src/ApplicationManager.cpp]:
```suggestion
#include "stp/new_plays/BallPlacement.h"
```

File: src/stp/new_plays/BallPlacement.cpp (hunk 5 above)
Comment (RobotJente) [inline src/stp/new_plays/BallPlacement.cpp]:
```suggestion
#include "stp/invariants/HaltGameStateInvariant.h"
#include "stp/invariants/WeHaveBallInvariant.h"
#include "stp/new_roles/BallPlacer.h"
```

File: src/stp/new_roles/BallPlacer.cpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
+//
+// Created by jesse on 07-04-20.
+//
+
+#include <include/roboteam_ai/stp/new_tactics/GetBall.h>
```
Comment (RobotJente) [inline src/stp/new_roles/BallPlacer.cpp]:
```suggestion
#include "stp/new_tactics/GetBall.h"
```

File: src/stp/new_roles/BallPlacer.cpp [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+#include "include/roboteam_ai/stp/new_roles/BallPlacer.h"
```
Comment (RobotJente) [inline src/stp/new_roles/BallPlacer.cpp]:
```suggestion
#include "stp/new_roles/BallPlacer.h"
```

File: src/stp/new_plays/BallPlacement.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    void BallPlacement::calculateInfoForRoles() noexcept {
+        if (stpInfos.find("ball_placer") != stpInfos.end()) {
+            auto ballTarget = Vector2(0,0);
+            stpInfos["ball_placer"].setPositionToMoveTo(ballTarget);
+        }
+        else {
+            RTT_ERROR("No ball placement robot assigned!")
+        }
+
```
Comment (RobotJente) [inline src/stp/new_plays/BallPlacement.cpp]:
```suggestion
```

File: src/ApplicationManager.cpp [hunk 10]
```diff
@@ -44,6 +45,7 @@ void ApplicationManager::start() {
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::Defend>());
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::DefensiveFormation>());
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::AggressiveFormation>());
+    plays.emplace_back(std::make_unique<rtt::ai::stp::play::BallPlacement>());
     playChecker.setPlays(plays);
```
Comment (Yuhanun) [inline src/ApplicationManager.cpp:49]: Why are these unique pointers again when they're owned by `plays` and you can just pass Play* instead?...

File: src/stp/new_plays/AggressiveFormation.cpp [hunk 11]
```diff
@@ -27,7 +27,7 @@ AggressiveFormation::AggressiveFormation() : Play() {
         std::make_unique<role::Formation>(role::Formation("offender_3"))};
 }
 
-uint8_t AggressiveFormation::score(world_new::World* world) noexcept { return 101; }
+uint8_t AggressiveFormation::score(world_new::World* world) noexcept { return 0; }
```
Comment (Yuhanun) [inline src/stp/new_plays/AggressiveFormation.cpp:30]: uh, 0?

**Review by Yuhanun — APPROVED** [review]

**Comment by Yuhanun** [comment]
LGTM apart from the score, assuming that is temporary

**Review by jordiluong — APPROVED** [review]

**Pull request merged**